dali/core/exec/tasking/scheduler.cc

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cassert>
+#include <mutex>
+#include <iostream>
+#include "dali/core/exec/tasking/scheduler.h"
+
+namespace dali::tasking {
+
+bool Scheduler::CheckTaskReady(SharedTask &task) noexcept {
+  assert(task->state <= TaskState::Pending);
+
+  for (auto &w : task->preconditions_)
+    if (!w->CheckComplete())
+      return false;
+  for (auto &w : task->preconditions_)
+    if (!w->TryAcquire(task)) {
+      // this should be a fatal error, but terminate and abort don't have messages
+      std::cerr
+          << "Internal error - resource acquisition failed for a resource known to be available"
+          << std::endl;
+      std::abort();
+    }
+
+  task->preconditions_.clear();
+  task->state = TaskState::Ready;
+  pending_.Remove(task);
+  ready_.push(std::move(task));
+  return true;
+}
+
+void Scheduler::Notify(Waitable *w) {
+  bool is_completion_event = dynamic_cast<CompletionEvent *>(w) != nullptr;
+  bool is_task = is_completion_event && dynamic_cast<Task *>(w);
```

Comment:
If Task is derived from CompletionEvent won't `is_completion_event` and `is_task` both be true or false at the same time?

Reply:
`is_task` implies `is_completion_event` but not vice versa - I envision other completion events (CUDA-related).

Reply:
Understood.